lib/IRGen/GenThunk.cpp

```diff
@@ -0,0 +1,98 @@
+//===--- GenDecl.cpp - IR Generation for Method Dispatch Thunks -----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements IR generation for class and protocol method dispatch
+//  thunks, which are used in resilient builds to hide vtable and witness table
+//  offsets from clients.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Callee.h"
+#include "Explosion.h"
+#include "GenDecl.h"
+#include "GenOpaque.h"
+#include "GenProto.h"
+#include "IRGenFunction.h"
+#include "IRGenModule.h"
+#include "ProtocolInfo.h"
+#include "Signature.h"
+#include "swift/IRGen/Linking.h"
+#include "llvm/IR/Function.h"
+
+using namespace swift;
+using namespace irgen;
+
+/// Find the entry point for a method dispatch thunk.
+llvm::Function *
+IRGenModule::getAddrOfDispatchThunk(SILDeclRef declRef,
+                                    ForDefinition_t forDefinition) {
+  LinkEntity entity = LinkEntity::forDispatchThunk(declRef);
+
+  llvm::Function *&entry = GlobalFuncs[entity];
+  if (entry) {
+    if (forDefinition) updateLinkageForDefinition(*this, entry, entity);
+    return entry;
+  }
+
+  auto &constantInfo = getSILModule().Types.getConstantInfo(declRef);
+  Signature signature = getSignature(constantInfo.SILFnType);
+  LinkInfo link = LinkInfo::get(*this, entity, forDefinition);
+
+  return createFunction(*this, link, signature);
+}
+
+static FunctionPointer lookupMethod(IRGenFunction &IGF,
+                                    SILDeclRef declRef) {
```

Comment:
Since you apparently intend to also handle class methods in this code, please name this function something protocol-related.

Reply:
I'm actually going to handle class methods inside lookupMethod().

Reply:
Okay.  My intuition would've been that there's not going to be much in common between the code paths, but if you can delegate all the FunctionPointer-building out of this file, then maybe all this has left is the swift between the two options.